File: comguanmuonehours/src/main/java/com/guanmu/onehours/fragment/FragmentVerifyCode.java
package com.guanmu.onehours.fragment;


import android.content.Context;
import android.os.Bundle;
import android.os.CountDownTimer;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import com.guanmu.onehours.R;
import com.guanmu.onehours.view.LoadingView;

/**
 * A simple {@link Fragment} subclass.
 */
public class FragmentVerifyCode extends BaseFragment implements View.OnClickListener {

    private OnFragmentInteractionListener mListener;
    private EditText mVerifyCodeView;

    private Button mRequestVerifyButton;
    private CountDownTimer countDownTimer;

    @Override
    public View createView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_verify_code, container, false);
        mVerifyCodeView = (EditText) view.findViewById(R.id.verify_code);

        mRequestVerifyButton = (Button) view.findViewById(R.id.request_verify_button);

        countDownTimer = new CountDownTimer(60000, 1000) {
            // A counter to allow user send verification code every 60s
            // Triggers onTick every 10s
            public void onTick(long millisUntilFinished) {
                mRequestVerifyButton.setText(String.format(getString(R.string.tip_wait_for_request), millisUntilFinished / 1000));
                mRequestVerifyButton.setEnabled(false);
            }

            public void onFinish() {
                mRequestVerifyButton.setText(getString(R.string.action_request_verify));
                mRequestVerifyButton.setEnabled(true);
            }
        };
        mRequestVerifyButton.setOnClickListener(this);
        countDownTimer.start();

        LoadingView verifyButton = (LoadingView) view.findViewById(R.id.verify_button);
        verifyButton.setOnClickListener(this);

        return view;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.verify_button:
                String code = mVerifyCodeView.getText().toString();

                // TODO
//                AVUser.verifyMobilePhoneInBackground(code, new AVMobilePhoneVerifyCallback() {
//                    @Override
//                    public void done(AVException e) {
//                        if (e == null) {
//                            success();
//                        } else {
//                            Toast.makeText(getActivity(), e.getLocalizedMessage(), Toast.LENGTH_LONG).show();
//                        }
//                    }
//                });
                break;
            case R.id.request_verify_button:
                mRequestVerifyButton.setText(R.string.tip_sending);
                mRequestVerifyButton.setEnabled(false);
//                AVUser.requestMobilePhoneVerifyInBackground(AVUser.getCurrentUser().getMobilePhoneNumber(), new RequestMobileCodeCallback() {
//                    @Override
//                    public void done(AVException e) {
//                        countDownTimer.start();
//                    }
//                });
        }
    }

    private void success() {
        if (mListener != null) {
            mListener.onFragmentSuccess();
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }


    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        void onFragmentSuccess();
    }
}

File: comguanmuonehours/src/main/java/com/guanmu/onehours/fragment/FragmentVerifyCode.kt
package com.guanmu.onehours.fragment


import android.content.Context
import android.os.Bundle
import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import androidx.fragment.app.Fragment
import com.guanmu.onehours.R
import com.guanmu.onehours.view.LoadingView

/**
 * A simple [Fragment] subclass.
 */
class FragmentVerifyCode : BaseFragment(), View.OnClickListener {

    private var mListener: OnFragmentInteractionListener? = null
    private var mVerifyCodeView: EditText? = null

    private var mRequestVerifyButton: Button? = null
    private var countDownTimer: CountDownTimer? = null

    public override fun createView(inflater: LayoutInflater, container: ViewGroup?,
                                   savedInstanceState: Bundle?): View {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_verify_code, container, false)
        mVerifyCodeView = view.findViewById<View>(R.id.verify_code) as EditText

        mRequestVerifyButton = view.findViewById<View>(R.id.request_verify_button) as Button

        countDownTimer = object : CountDownTimer(60000, 1000) {
            // A counter to allow user send verification code every 60s
            // Triggers onTick every 10s
            override fun onTick(millisUntilFinished: Long) {
                mRequestVerifyButton!!.text = String.format(getString(R.string.tip_wait_for_request), millisUntilFinished / 1000)
                mRequestVerifyButton!!.isEnabled = false
            }

            override fun onFinish() {
                mRequestVerifyButton!!.text = getString(R.string.action_request_verify)
                mRequestVerifyButton!!.isEnabled = true
            }
        }
        mRequestVerifyButton!!.setOnClickListener(this)
        countDownTimer!!.start()

        val verifyButton = view.findViewById<View>(R.id.verify_button) as LoadingView
        verifyButton.setOnClickListener(this)

        return view
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.verify_button -> {
                val code = mVerifyCodeView!!.text.toString()
            }
            R.id.request_verify_button -> {
                mRequestVerifyButton!!.setText(R.string.tip_sending)
                mRequestVerifyButton!!.isEnabled = false
            }
        }// TODO
        //                AVUser.verifyMobilePhoneInBackground(code, new AVMobilePhoneVerifyCallback() {
        //                    @Override
        //                    public void done(AVException e) {
        //                        if (e == null) {
        //                            success();
        //                        } else {
        //                            Toast.makeText(getActivity(), e.getLocalizedMessage(), Toast.LENGTH_LONG).show();
        //                        }
        //                    }
        //                });
        //                AVUser.requestMobilePhoneVerifyInBackground(AVUser.getCurrentUser().getMobilePhoneNumber(), new RequestMobileCodeCallback() {
        //                    @Override
        //                    public void done(AVException e) {
        //                        countDownTimer.start();
        //                    }
        //                });
    }

    private fun success() {
        if (mListener != null) {
            mListener!!.onFragmentSuccess()
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }


    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        fun onFragmentSuccess()
    }
}
